lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,28 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
```

Comment:
> I was thinking of an implicit subject. 

Good point...I forgot about that case here...but yeah, that's definitely a concern.  One potential "middle ground" is to do the following:

* Move the `ReentrantMutex` out of `Memoized` and into `MemoizedHelpers`, and instantiate that when the example group is instantiated.
* Rather than doing `__memoized.fetch_or_store` everywhere, refactor so that there's a `__fetch_or_store_memoized` method defined in `MemoizedHelpers`, with an implementation like:

``` ruby
def __fetch_or_store_memoized(key)
  @__memoized_mutex.synchronize do
    @__memoized ||= {}
    @__memoized.fetch(key) { @__memoized[key] = yield }
  end
end
```

That way, both the hash instantiation and the hash access are protected by the same mutex.  Not sure if it would perform better in the case where none of the memoized helpers are used, but it's worth considering.